packages/block-editor/src/components/default-style-picker/index.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ */
+import { get } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useCallback } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { SelectControl } from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+
+export default function DefaultStylePicker( { blockName } ) {
+	const {
+		defaultStyle,
+		onUpdateDefaultBlockStyles,
+		styles,
+	} = useSelect(
+		( select ) => {
+			const settings = select( 'core/block-editor' ).getSettings();
+			return {
+				defaultStyle: get( settings, [ 'defaultBlockStyles', blockName ] ),
+				onUpdateDefaultBlockStyles: settings.onUpdateDefaultBlockStyles,
```

Comment:
This to me is quite strange. Have we discussed the APIs to set user preferences and exposing them in Settings like this? Otherwise I'd expect this change to be applied in core components by calling our dispatcher API.

Reply:
Hi @mcsf,

> Have we discussed the APIs to set user preferences and exposing them in Settings like this? Otherwise I'd expect this change to be applied in core components by calling our dispatcher API.

I think this was not discussed before. The edit-post module is handling the block style preference, the UI to change the style is part of the style picker (block editor module). The block editor module should not be aware of the edit post module, so it can not use our dispatcher API to update the preference. Keeping this abstraction allows users of the block editor module that may not use our data module to take advantage of the default style changer UI easily.
We have not yet a case similar to this one, but passing functions as settings to bock editor is not something new. For example, we pass the function that allows media uploads as a setting. The cases are not that different as both use a function that allows the block editor to change some state external to it (add a media file to the WordPress media lib for media uploads, change default block style and save on local storage for default style change).

Another option would be passing onUpdateDefaultBlockStyles as an optional property of the BlockEditor component, but that increases the complexity even more as we would need to use context to pass this property does down to the UI components that apply the change.

Reply:
Thanks for the reply, @jorgefilipecosta.

> The edit-post module is handling the block style preference, the UI to change the style is part of the style picker (block editor module). The block editor module should not be aware of the edit post module, so it can not use our dispatcher API to update the preference. Keeping this abstraction allows users of the block editor module that may not use our data module to take advantage of the default style changer UI easily.

So in my mind we'd use some of our Data features such as controls and registry awareness so that `DefaultStylePicker` only interacts with `block-editor` data, in particular by dispatching through `block-editor`, while behind the scenes the `block-editor` store would bridge with `editor`. The key here would be to not make assumptions about unicity of those stores, and I'm not familiar enough with the current state of our data layer to say how feasible this is. @youknowriad, do you know?

> We have not yet a case similar to this one, but passing functions as settings to bock editor is not something new. For example, we pass the function that allows media uploads as a setting.

I had a quick glance at the other objects exposed as settings and failed to spot those functions — my bad.

> Another option would be passing onUpdateDefaultBlockStyles as an optional property of the BlockEditor component, but that increases the complexity even more as we would need to use context to pass this property does down to the UI components that apply the change.

Yeah, probably not worth it.

Taking a step back, I think Settings are a bit of chimera, in that they are "just" store data that's passed around but it's made available through separate channels, usually premised on remembering users' preferences. That is, it's the preferences themselves which are the settings, not the functions by which preferences are set. That's why it didn't seem right to me to expose the functions. That said, if we have precedents for this then I don't want to keep this PR from moving. :)

Reply:
I don't think we have a good answer here tbh. Some thoughts though: 

> So in my mind we'd use some of our Data features such as controls and registry awareness so that DefaultStylePicker only interacts with block-editor data, in particular by dispatching through block-editor, while behind the scenes the block-editor store would bridge with editor. The key here would be to not make assumptions about unicity of those stores, and I'm not familiar enough with the current state of our data layer to say how feasible this is. @youknowriad, do you know?

There's no straight-forward API to do that at the moment. It might be possible to achieve by duplicating the data in two stores (subcribing and redispatching on change) but it feels wrong. That said, I'm not certain that's the perfect API for that use case.

---
Thinking of the `BlockEditor` usage outside of the WordPress context help us think about the best API here I think.

So far, we have four main ways to "extend/use" the block editor: settings, composing UI components, slots and filters. In my opinion, from a third-party developer's perspective, the best devX would be:

 - Compose my own UI
 - Provide settings
 - Slots comes after as while it's useful sometimes, if I'm able to compose my own UI, I shouldn't need them as much.
 - Filters for very specific numbers/strings... we want to be tweakable

In this particular use-case, my first question is how valuable this feature/setting would be outside of WordPress? If it's something users would want to tweak often, I think the "settings" might be the best place for it. `mediaUpload` is a good example of something that any third party editor might want to tweak while keeping the same blocks...

It doesn't feel like something too common for me so I'd personally try to find the less impacting the `block-editor` module. It doesn't mean I won't use `settings` for it though especially if I can find a good "setting property" for it. It looks like it's composed of two settings: `the value of the settings` and a `change handler`. Also how can we disable it/enable it, is it by the presence of the handler. What if it's something like:

```
preferedStyleVariations: {
     value,
     onChange
}
```

and the presence of the whole object indicate support or not?

---
I'm still not convinced totally yet though :) I think the main problem we're facing here is that the `BlockEditorProvider` component is thought to be a component that edits one value: The blocks. So it has a  `value`  prop for the blocks and an `onChange` handler. And the settings are just a way to tweak its behavior. In this Particular PR we're making the block editor edit a secondary "value", the preferred styles.

Reply:
sorry for the long comment with no real solution :)

Reply:
Thanks, I think that helps put things in perspective.

Since there's no solid path forward, are we OK with proceeding with the current solution?

Reply:
I'm personally fine with it. We should just have this UI disabled by default if you use the block editor without these settings. We could consider marking it experimental?

Aside: Do you think it's possible to build this as a Slot/UI component in the `editor module` instead? Conceptually speaking, this would mean we're just editing a data of the post editor but not the block editor, we're just trying to "merge" the UI. Did we give any thoughts to that possibility?

Reply:
> I'm personally fine with it. We should just have this UI disabled by default if you use the block editor without these settings. We could consider marking it experimental?

Yes, let's make it experimental.

> Aside: Do you think it's possible to build this as a Slot/UI component in the `editor module` instead? Conceptually speaking, this would mean we're just editing a data of the post editor but not the block editor, we're just trying to "merge" the UI. Did we give any thoughts to that possibility?

That sounds pretty interesting, I'd like to see it tried.

@jorgefilipecosta, what do you think of first moving forward by marking the exposed bits from this PR as experimental? Later, do you think it's worth exploring the `editor`-side Slot/Fill approach?

Reply:
Hi @youknowriad, @mcsf,
I updated the code marked it as experimental and followed @youknowriad suggestion:
```
preferedStyleVariations: {
     value,
     onChange
}
```

I thought about the Slot Fill approach, basically, add a slot to the styles panel, but that slot seemed arbitrary. Why it only appears in the styles panel inside block inspector and not in styles inside block switcher? Why we have a slot in this panel but not other panels?
We may have cases of multiple block editors, how do I enable style change functionality in some of the block editors but not others?
This approach allows us to pass different change handlers per block editor the slot fill seemed more limited and more complex.